### PR TITLE
Upgraded needToChargeBatteries() and driveRoverToTarget()

### PR DESCRIPTION
**Pull request by william64644**
needToChargeBatteries() is now following the clean code rules bellow:
Functions should not do more than one thing.
Functions must never do more than what they imply to do.

Clean code rules and best practices in general improve code scalability, legibility and reliability.

The process and the changes were decently documented along the commits.

### CODE REVIEW

File: src/com/pesterenan/controllers/RoverController.java [hunk 1]
```diff
@@ -1,360 +1,395 @@
...
+			steeringLine = drawing.addDirection(roverDirection.toTriplet(), pontoRefRover, 10, true);
+			steeringLine.setColor(new Triplet<>(1.0, 0.0, 1.0));
+			for (int i = 0; i < MAX_RADAR_LINES; i++) {
+				Drawing.Line line = drawing.addDirection(roverDirection.toTriplet(), pontoRefSuperficie, 1, true);
+				line.setColor(new Triplet<>(0.0, (1.0 / MAX_RADAR_LINES), 0.0));
+				line.setThickness(0.2f);
+				radarLines.add(line);
+			}
+			pathFinding = new PathFinding(getConexao());
+			haveSolarPanels = detectSolarPanels();
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:82]: This shouldn't be a variable that is updated only here, we need to check this before every recharge cycle, as the rover could've lost some solar panel.

File: src/com/pesterenan/controllers/RoverController.java [hunk 2]
```diff
@@ -1,360 +1,395 @@
...
+			StatusJPanel.setStatus("Segundos de Carga: " + chargeTime);
+			if (chargeTime < 1 || chargeTime > 21600) {
+				chargeTime = 3600;
+			}
+			centroEspacial.warpTo((centroEspacial.getUT() + chargeTime), 10000, 4);
+			naveAtual.getControl().setLights(true);
+		}
+	}
+	
+	private boolean detectSolarPanels() throws RPCException {
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:214]: Nice, abstracting away this function is the right thing to do. As it already returns a boolean, we can call it "hasWorkingSolarPanels" and call it whenever we start to recharge the rover.

File: src/com/pesterenan/controllers/RoverController.java [hunk 3]
```diff
@@ -1,360 +1,395 @@
...
+					naveAtual.getControl().setBrakes(true);
+					pathFinding.removePathsCurrentPoint();
+					if (commands.get(Modulos.TIPO_ALVO_ROVER.get()).equals(Modulos.MARCADOR_MAPA.get()) &&
+							pathFinding.isPathToTargetEmpty()) {
+						pathFinding.removeWaypointFromList();
+						pathFinding.findNearestWaypoint();
+					}
+				}
+			} else { // Rover needs recharging 
+				if (haveSolarPanels) {
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:145]: You can call the "hasWorkingSolarPanels" function right here.

File: src/com/pesterenan/controllers/RoverController.java [hunk 4]
```diff
@@ -1,360 +1,395 @@
...
+			double chargeTime = 0;
+			double TotalEnergyFlow = 0;
+			List<SolarPanel> solarPanels = naveAtual.getParts()
+			                                        .getSolarPanels()
+			                                        .stream()
+			                                        .filter(RoverController::isSolarPanelNotBroken)
+			                                        .collect(Collectors.toList());
+
+			for (SolarPanel sp : solarPanels) {
+				TotalEnergyFlow += sp.getEnergyFlow();
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:202]: Use camelCase when naming variables.

File: src/com/pesterenan/controllers/RoverController.java [hunk 5]
```diff
@@ -1,360 +1,395 @@
...
+	private ReferenceFrame pontoRefRover;
+	private boolean isAutoRoverRunning = true;
+	private Drawing drawing;
+	private Stream<Float> bateriaAtual;
+	private Drawing.Line steeringLine;
+	private PathFinding pathFinding;
+	private Vector targetPoint = new Vector();
+	private Vector roverDirection = new Vector();
+	private Drawing.Line dirRover;
+	private boolean haveSolarPanels;
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:44]: We won't need this field.

File: src/com/pesterenan/controllers/RoverController.java [hunk 6]
```diff
@@ -1,360 +1,395 @@
...
+			} else {
+				setNextPointInPath();
+			}
+			
+			
+			if (!needToChargeBatteries()) {
+				if (isFarFromTarget()) {
+					naveAtual.getControl().setBrakes(false);
+					driveRover();
+				} else { // Rover arrived at destiny
```
Comment (Pesterenan) [inline src/com/pesterenan/controllers/RoverController.java:135]: This should be abstracted away as a function, but I'll take care of that later.

**Review by Pesterenan — CHANGES_REQUESTED** [review]
These are the changes that I saw for now. Also there's some blank lines here and there. Also try to pull other changes to the dev branch! Thanks for the help!

File: src/com/pesterenan/controllers/RoverController.java (hunk 1 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java]: haveSolarPanels don't exist anymore, instead, hasWorkingSolarPanels() is called dynamically

File: src/com/pesterenan/controllers/RoverController.java (hunk 2 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java]: Renaming detectSolarPanels() to hasWorkingSolarPanels() done

File: src/com/pesterenan/controllers/RoverController.java (hunk 3 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java]: Now hasWorkingSolarPanels() is being called here

File: src/com/pesterenan/controllers/RoverController.java (hunk 4 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java]: Naming convention is now fixed

File: src/com/pesterenan/controllers/RoverController.java (hunk 5 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java]: haveSolarPanels is not being used anymore and got deleted

File: src/com/pesterenan/controllers/RoverController.java (hunk 6 above)
Comment (william64644) [reply src/com/pesterenan/controllers/RoverController.java:132]: Sounds like a good choice, as I don't really know how to manipulate your GUI

**Comment by william64644** [comment]
PS: You can talk to me in Portuguese if you prefer, I'm brazilian, I just like to keep the English as a default in my workflow

**Comment by william64644** [comment]
I tried switching the base of my branch to your Dev branch but it said there was a conflict and it wasn't even clear what the problem was for me to fix it

**Comment by Pesterenan** [comment]
Boa William, é nóis mano! Gostei das mudanças, vou ajustar os arquivos de configuração, assim você não vai precisar ficar trocando no fim. Depois disso, vou atualizar tanto a Dev quanto a Master, aí quando vc for subir código, pega da master, sobe uma branch, e depois a gente faz um merge na dev primeiro, blzinha?

**Review by Pesterenan — APPROVED** [review]

**Pull request merged**